### PR TITLE
Convert API layer error responses into Errors that conform to the API spec

### LINKED ISSUE #883: Implement new error structure in backend.
## Problem

We've agreed on a standard error structure for the API in #560. Now we need to implement it.

## Proposed solution

I'll format these in terms of desired outcomes. Whoever works on this can feel free to split these up into separate issues if that's helpful.
- All current API error responses should follow our error structure.
- All error codes and their corresponding `details` fields are documented somewhere.
- We have a standard way of generating new error codes.

## Additional context

- #560 
- [API Standards wiki page](https://wiki.mathesar.org/en/engineering/architecture/api-standards)

### PR DESCRIPTION
**Pull request by silentninja**
Fixes #883 

Transforms errors thrown at the API layer into proper error response that conform to the API spec
 
**Technical details**
1. Converts existing Exceptions and Errors into proper APIException with the exception detail structure that conforms to the API spec by wrapping the Exceptions into adapter Exception 
2. Adds Mixins to Serializers class to change validation error structure to include details necessary to conform to the API error spec

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by silentninja** [comment]
Blocked by #986 and #990

File: mathesar/exception_handlers.py
Comment (mathemancer) [inline mathesar/exception_handlers.py:6]: Is something missing here, or am I not understanding?

**Review by mathemancer — CHANGES_REQUESTED** [review]
Overall, I think this is a major improvement.  It's easy to see where and how to extend, and I think this will help us easily keep our error codes organized.  Great work.

There's one spot I noticed where either I don't understand what's happening, or there's something missing.  Which is it?  

Also, I suggest waiting until @kgodey has a chance to at least look at this, since she's more familiar with idiomatic Django error handling than I am.

File: mathesar/exception_handlers.py
Comment (silentninja) [reply mathesar/exception_handlers.py:6]: I split error handling into multiple PR as I felt the original PR was huge, that is the reason for so many incomplete placeholders. This is supposed to be a last resort conversion layer which would map any non spec Exception into its equivalent ApiException, l will open a PR that would make use of this.

**Comment by silentninja** [comment]
I am not really sure on how to organise the error codes and I would like to make sure if the naming convention for `Errors` actually make sense.

**Comment by kgodey** [comment]
I'll wait to review this until it's unblocked and the merge conflicts are fixed. The PRs which were blocking it are both merged, so hopefully it's ready to review soon.

File: mathesar/api/exceptions/exceptions.py
Comment (mathemancer) [inline mathesar/api/exceptions/exceptions.py:17]: I prefer this style:
```suggestion
def get_default_exception_detail(
        exception,
        error_code=ErrorCodes.UnknownError.value,
        message=None,
        field=None,
        details=None,
):
```

It makes it easier to pick out all the arguments, as well as modify them (adding, reordering, removing).

File: mathesar/api/serializers/tables.py
Comment (mathemancer) [inline mathesar/api/serializers/tables.py:7]: I think either splitting up these imports or (even better) using parentheses would be better.  Escaped line breaks aren't usually needed in Python.

File: mathesar/api/serializers/tables.py
Comment (mathemancer) [inline mathesar/api/serializers/tables.py:92]:
```suggestion
                raise DuplicateTableAPIException(
                    e,
                    message=f"Relation {validated_data['name']} already exists in schema {schema.id}",
                    field="name",
                    status_code=status.HTTP_400_BAD_REQUEST,
                )
```
This has the added benefit of making it more reasonable to avoid splitting the message across two lines.

File: mathesar/api/viewsets/columns.py
Comment (mathemancer) [inline mathesar/api/viewsets/columns.py:27]:
```suggestion
        column_attnum_list = [
            result[0] for result in
            get_columns_attnum_from_names(table.oid, sa_column_name, table.schema._sa_engine)
        ]
```
This one's quite subjective, but the suggested style is more aligned with the rest of the list comprehensions in the repo.

**Review by mathemancer — CHANGES_REQUESTED** [review]
I think this looks pretty good, functionality-wise (subject to @kgodey 's thoughts.  Good work on that.  

I made some formatting suggestions.  These are meant to be examples, not comprehensive.  I think applying similar transformations to most function argument sets would help this PR align with other parts of the code base better.  Also, while we're not being to fussy about line length, there are a number of places (e.g., comments, docstrings) where there's not much of a reason to exceed the PEP8 specified 79.

File: mathesar/api/exceptions/error_codes.py
Comment (kgodey) [inline mathesar/api/exceptions/error_codes.py:7]: `API`, not `Api`

File: mathesar/api/exceptions/error_codes.py
Comment (kgodey) [inline mathesar/api/exceptions/error_codes.py:16]: Could you use `DB` here?

File: mathesar/api/exceptions/mixins.py
Comment (kgodey) [inline mathesar/api/exceptions/mixins.py:25]: Please remove blank line

File: mathesar/api/exceptions/mixins.py
Comment (kgodey) [inline mathesar/api/exceptions/mixins.py:81]: If you're splitting a list or tuple into multiple lines, I much prefer this pattern to having a few listed on every line:
```
 'string': [
    'CharField',
    'EmailField',
    'RegexField',
    'SlugField',
    'URLField',
    'UUIDField',
    'FilePathField',
]

Comment (kgodey) [reply mathesar/api/exceptions/mixins.py:81]: Also applies to everywhere else.

File: mathesar/api/exceptions/mixins.py
Comment (kgodey) [inline mathesar/api/exceptions/mixins.py:115]: If you're splitting a dict into different lines, I prefer this:

```python
kwargs.update({
    'max_length': getattr(field, 'max_length', None),
    'min_length': getattr(field, 'min_length', None),
    'value': field_data
})
```

Also applies to other instances

File: mathesar/api/exceptions/mixins.py
Comment (kgodey) [inline mathesar/api/exceptions/mixins.py:80]: Can't you read the existing field map and only make changes to what's necessary rather than redoing the whole thing?

File: mathesar/api/exceptions/mixins.py
Comment (kgodey) [inline mathesar/api/exceptions/mixins.py:99]: Again, can you reuse as much of the original method as possible? If the upstream package made improvements or fixes, it would be nice for us to reuse that as much as possible.

File: mathesar/api/viewsets/columns.py
Comment (kgodey) [inline mathesar/api/viewsets/columns.py:56]: I prefer be something like:

```python
                    raise exceptions.DuplicateTableAPIException(
                        e,
                        message=f'Column {name} already exists',
                        field='name',
                        status_code=status.HTTP_400_BAD_REQUEST
                    )
```

Comment (kgodey) [reply mathesar/api/viewsets/columns.py:56]: Also applies to the rest.

File: mathesar/exception_handlers.py
Comment (kgodey) [inline mathesar/exception_handlers.py:9]: I think `prettify_response` would be a better name here.

**Review by kgodey — CHANGES_REQUESTED** [review]
@silentninja Looks good overall, very nice work. Most of my comments are style related.

File: mathesar/exception_handlers.py
Comment (silentninja) [reply mathesar/exception_handlers.py:9]: There are two types of response data which are not formatted correctly.
1. Response data which does not have a proper structure
2,  Response data which has proper structure but has some quirks due to `drf` way of handling exceptions.

In case of (2) we are fixing the pretty error. Moreover we are checking if the error is pretty in previous condition block, so it would be conflicting to name it `pretty`,

File: mathesar/api/exceptions/exceptions.py
Comment (silentninja) [reply mathesar/api/exceptions/exceptions.py]: Should the arguments indent be 4 spaces or 8 spaces.

```

def get_default_exception_detail(
        exception,

```

Comment (mathemancer) [reply mathesar/api/exceptions/exceptions.py]: 8 spaces is preferred, since it makes it more visually distinct from the function body.

File: mathesar/api/exceptions/exceptions.py
Comment (kgodey) [inline mathesar/api/exceptions/exceptions.py:82]: Do we need to specify arguments to exception classes when they're `None`? Can't you use `**kwargs` and handle `message`, `field`, and `details` in the `__init__` method of the `MathesarAPIException` class? That will reduce boilerplate code for new exceptions.

File: config/settings.py
Comment (kgodey) [inline config/settings.py:185]: You missed the one line per item formatting here.

File: mathesar/api/exceptions/exceptions.py
Comment (kgodey) [inline mathesar/api/exceptions/exceptions.py:35]: Since this is the base class, I think it would make sense to move this to a `base.py` file, editing this exception will affect everything else.

File: mathesar/api/exceptions/mixins.py
Comment (kgodey) [reply mathesar/api/exceptions/mixins.py:83]: Okay, thanks. Please do that and also create an issue in our repo to replace our code with the library's code ASAP.

**Review by kgodey — CHANGES_REQUESTED** [review]
@silentninja Looks mostly good, made some more suggestions to make the exception code more DRY and improve organization and formatting.

File: mathesar/exception_handlers.py
Comment (mathemancer) [inline mathesar/exception_handlers.py:37]: I'm not sure what this means.  Consider a more descriptive function name instead of `fix_error_response`.

**Review by mathemancer — CHANGES_REQUESTED** [review]
Overall, much more readable for me.  The functionality seems to be in place.  I made a couple of small requests.

File: mathesar/api/exceptions/exceptions.py
Comment (silentninja) [reply mathesar/api/exceptions/exceptions.py]: Moved the status code out of `init`. Some Exceptions have `status_code` in their `init` so that it can be overridden. I refactored and moved the exceptions into proper modules, so that consistency can be maintained.

File: mathesar/api/exceptions/exceptions.py
Comment (silentninja) [reply mathesar/api/exceptions/exceptions.py]: This would make the error code tightly coupled with an exception class also making it hard to name the error code properly.

File: mathesar/api/exceptions/exceptions.py
Comment (silentninja) [reply mathesar/api/exceptions/exceptions.py]: It is nice to see what arguments are accepted by a method, rather than having to check the docs or inspect the method. I moved certain fields which don't have to be overridden outside of the `init` method.

File: mathesar/api/exceptions/exceptions.py
Comment (silentninja) [reply mathesar/api/exceptions/exceptions.py]: I refactored exceptions.py and moved the expectations into appropriate modules

File: mathesar/exception_handlers.py
Comment (kgodey) [reply mathesar/exception_handlers.py:9]: Can you name it `standardize_error_response` then? I think it's more descriptive.

**Review by kgodey — APPROVED** [review]
Since @mathemancer is out for the next couple of days, I don't think we should wait for him to re-review this before merge. I did look through his comments and verified they're addressed.

I added one minor comment, please fix it and merge yourself @silentninja.

**Pull request merged**
